packages/contracts/contracts/QuadraticVoteCollectModule.sol

```diff
@@ -0,0 +1,260 @@
+            IERC20(_currency).safeTransfer(treasury, treasuryAmount);
+        }
+    }
+
+    function _processCollectWithReferral(
+        uint256 referrerProfileId,
+        address collector,
+        uint256 profileId,
+        uint256 pubId,
+        bytes calldata data
+    ) internal {
+        /// decode  data
+        (, uint256 _amount, VotingData memory votingData) = abi.decode(data, (address, uint256, VotingData));
+        require(
+            block.timestamp > votingData.roundStartTime && block.timestamp < votingData.roundEndTime,
+            "Round is not in session"
+        );
+
+        //TODO check signature validity
+
+        //validate data
+        uint256 amount = _dataByPublicationByProfile[profileId][pubId].amount;
+        address currency = _dataByPublicationByProfile[profileId][pubId].currency;
+        uint256 referralFee = _dataByPublicationByProfile[profileId][pubId].referralFee;
+
+        _validateDataIsExpected(data, currency, amount);
+
+        //transfer erc20 from collector to this contract
+        IERC20(currency).safeTransferFrom(collector, address(this), _amount);
+
+        // address recipient = _dataByPublicationByProfile[profileId][pubId].recipient;
+        address treasury;
+        uint256 treasuryAmount;
+        uint256 adjustedAmount;
+        uint256 referralAmount;
+
+        if (referralFee != 0) {
+            // Avoids stack too deep
+            {
+                uint16 treasuryFee;
+                (treasury, treasuryFee) = _treasuryData();
+                treasuryAmount = (amount * treasuryFee) / BPS_MAX;
+                referralAmount = (adjustedAmount * referralFee) / BPS_MAX;
+                adjustedAmount = amount - treasuryAmount - referralAmount;
+            }
+            // The reason we levy the referral fee on the adjusted amount is so that referral fees
+            // don't bypass the treasury fee, in essence referrals pay their fair share to the treasury.
+            _processReferral(referrerProfileId, referralAmount, currency);
+        }
+        //address recipient = _dataByPublicationByProfile[profileId][pubId].recipient;
+
+        if (treasuryAmount != 0) {
+            IERC20(currency).safeTransfer(treasury, treasuryAmount);
+        }
+
+        //cast vote
+        _vote(adjustedAmount, data);
+    }
+
+    function _vote(uint256 adjustedAmount, bytes calldata data) internal {
+        (address _currency, , VotingData memory votingData) = abi.decode(data, (address, uint256, VotingData));
+
+        // encode vote
+        bytes memory vote = abi.encode(_currency, adjustedAmount, votingData.grantsRoundAddress);
+
+        /// declare votes array
+        bytes[] memory votes = new bytes[](1);
+
+        /// cast vote into array because that's how gitcoin likes it.
+        votes[0] = vote;
+
+        // approve voting strategy to spend erc20
+        IERC20(_currency).approve(votingData.votingStrategyAddress, adjustedAmount);
```

Comment:
Would we need a approval for every executed vote?

Reply:
Yes, that's how it's set up now.  Probably worth re-thinking now that we are looking at implementing claps.

Reply:
Even without the claps, modules are set with allowance so this shouldn't be part of the voting flow if I get the modules of Lens correctly: https://docs.lens.xyz/docs/approved-allowance-of-modules

Reply:
Approve and vote in same tx